### PR TITLE
feat(tsdb): Enable marking clusters as non-durable for writes

### PR DESCRIPTION
**Pull request by tkaemming**
This prevents failing to write to these clusters from failing the entire write operation.

### CODE REVIEW

File: src/sentry/tsdb/redis.py [hunk 1]
```diff
@@ -189,8 +224,13 @@ def incr_multi(self, items, timestamp=None, count=1, environment_id=None):
...
 
-        for cluster, environment_ids in self.get_cluster_groups(set([None, environment_id])):
-            with cluster.map() as client:
+        for (cluster, durable), environment_ids in self.get_cluster_groups(
+                set([None, environment_id])):
+            manager = cluster.map()
+            if not durable:
+                manager = SuppressionWrapper(manager)
+
+            with manager as client:
```
Comment (tkaemming) [inline src/sentry/tsdb/redis.py]: This works on the assumption that all potentially problematic IO is deferred to the `__exit__` method of the wrapped context manager. I think this is correct but should probably validate this to be sure.

If there are exceptions raised within the context manager (for example, a bug in the code that results in a `TypeError`), those will be propagated. I personally think this is a good thing, but open to other perspectives.

File: src/sentry/tsdb/redis.py [hunk 2]
```diff
@@ -817,9 +883,17 @@ def merge_frequencies(self, model, destination, sources, timestamp=None, environ
                             ]
                         )
 
+            try:
+                responses = cluster.execute_commands(exports)
+            except Exception:
+                if durable:
+                    raise
+                else:
+                    continue
```
Comment (tkaemming) [inline src/sentry/tsdb/redis.py]: This control flow is pretty gross, but I wanted to minimize the scope of this change and this was the least invasive way I could think to do so.

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/tsdb/redis.py [hunk 3]
```diff
@@ -37,6 +37,32 @@
...
+
+    def __exit__(self, *args):
+        if any(arg is not None for arg in args):
+            # propagate exception raised in managed context
+            return
+
+        try:
+            return self.wrapped.__exit__(*args)
+        except self.types:
+            # suppress exception raised when exiting wrapped manager
```
Comment (tkaemming) [inline src/sentry/tsdb/redis.py]: I'd personally like to log this but I feel like that'd result in a compounding problem…?

**Review by tkaemming — COMMENTED** [review]
The rationale behind this change is that it allows us to prevent non-essential writes to things like the new environment cluster from affecting callers that don't handle TSDB write failures gracefully (and probably shouldn't have to.)

**Comment by tkaemming** [comment]
(Glances in the general direction of @JTCunning.)

File: src/sentry/tsdb/redis.py [hunk 4]
```diff
@@ -580,7 +638,7 @@ def delete_distinct_counts(self, models, keys, start=None, end=None,
 
         rollups = self.get_active_series(start, end, timestamp)
 
-        for cluster, environment_ids in self.get_cluster_groups(environment_ids):
+        for (cluster, duraable), environment_ids in self.get_cluster_groups(environment_ids):
```
Comment (tkaemming) [inline src/sentry/tsdb/redis.py]: Dammit

File: src/sentry/tsdb/redis.py (hunk 1 above)
Comment (tkaemming) [reply src/sentry/tsdb/redis.py]: Also, if this isn't the case, I guess a fallback is to just always prevent the exception from propagating as part of the suppression manager, but widely increasing the scope of that suppression makes me feel pretty bad.

Comment (tkaemming) [reply src/sentry/tsdb/redis.py]: Well then: executing *any* command within the context manager results in a connection being established via the `CommandBuffer` constructor: https://github.com/getsentry/rb/blob/569d1d13311f6c04bae537fc17e75da430e4ec45/rb/clients.py#L90

Full traceback in the expandable details: <details>

```
<ipython-input-6-3d9d22111bf0> in <module>()
      1 with cluster.map() as c:
----> 2     c.get('k')
      3 

/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/redis/client.pyc in get(self, name)
    878         Return the value at key ``name``, or None if the key doesn't exist
    879         """
--> 880         return self.execute_command('GET', name)
    881 
    882     def __getitem__(self, name):

/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/rb/clients.pyc in execute_command(self, *args, **options)
    331         router = self.connection_pool.cluster.get_router()
    332         host_id = router.get_host_for_command(args[0], args[1:])
--> 333         buf = self._get_command_buffer(host_id, args[0])
    334         return buf.enqueue_command(args[0], args[1:], options)
    335 

/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/rb/clients.pyc in _get_command_buffer(self, host_id, command_name)
    349             return self.connection_pool.get_connection(
    350                 command_name, shard_hint=host_id)
--> 351         buf = CommandBuffer(host_id, connect, self.auto_batch)
    352         self._cb_poll.register(host_id, buf)
    353         return buf

/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/rb/clients.pyc in __init__(self, host_id, connect, auto_batch)
     88         self.connection = None
     89         self._connect_func = connect
---> 90         self.connect()
     91         self.commands = []
     92         self.pending_responses = []

/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/rb/clients.pyc in connect(self)
    107         # Ensure we're connected.  Without this, we won't have a socket
    108         # we can select over.
--> 109         self.connection.connect()
    110 
    111     def reconnect(self):

/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/redis/connection.pyc in connect(self)
    437             return
    438         try:
--> 439             sock = self._connect()
    440         except socket.error:
    441             e = sys.exc_info()[1]

/Users/ted/.virtualenvs/sentry/lib/python2.7/site-packages/redis/connection.pyc in _connect(self)
    480 
    481                 # connect
--> 482                 sock.connect(socket_address)
    483 
    484                 # set the socket_timeout now that we're connected

/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/socket.pyc in meth(name, self, *args)
    226 
    227 def meth(name,self,*args):
--> 228     return getattr(self._sock,name)(*args)
    229 
    230 for _m in _socketmethods:
```
</details>

Comment (tkaemming) [reply src/sentry/tsdb/redis.py]: 32f1efe1d6f64b19ea018391980265413b80e436 now prevents any exception raised within the managed block or exiting of the wrapped manager from being propagated.

This is a super hamfisted approach to handling this but I don't see a better alternative. 😞

Comment (mattrobenolt) [reply src/sentry/tsdb/redis.py]: 🍔 👊

Comment (mattrobenolt) [reply src/sentry/tsdb/redis.py]: :pig: 👊

Comment (JTCunning) [reply src/sentry/tsdb/redis.py]: 💻 💪 💪

**Review by mattrobenolt — APPROVED** [review]
wcgw

**Comment by tkaemming** [comment]
> wcgw

Nothing, technically

**Pull request merged**